File: player-cli/internal/engine/engine_test.go
package engine_test

import (
	"strings"
	"testing"

	"zaistory/player-cli/internal/engine"
	"zaistory/player-cli/internal/ir"
)

// fakeUI gioca una sequenza di token predefinita e registra quello che vede.
// Serve a testare l'engine senza terminale: e' la stessa idea dello script di
// playthrough, ridotta al minimo.
type fakeUI struct {
	toks     []string
	pos      int
	problems []string
	log      []string
	hidden   map[string]string // azione -> motivo per cui era nascosta
}

func newFake(toks ...string) *fakeUI {
	return &fakeUI{toks: toks, hidden: map[string]string{}}
}

func (f *fakeUI) next() (string, bool) {
	if f.pos >= len(f.toks) {
		return "", false
	}
	t := f.toks[f.pos]
	f.pos++
	return t, true
}

func (f *fakeUI) SceneEnter(st *engine.State, sc *ir.Scene) { f.log = append(f.log, "scena:"+sc.ID) }
func (f *fakeUI) Beat(sc *ir.Scene, b ir.NarrationBeat, i, n int) error {
	f.log = append(f.log, "beat:"+sc.ID)
	return nil
}
func (f *fakeUI) Line(sc *ir.Scene, nodeID string, n *ir.DialogueNode) error {
	f.log = append(f.log, "nodo:"+nodeID)
	return nil
}
func (f *fakeUI) Narration(text string, v *ir.VoiceSpec) { f.log = append(f.log, "narr") }
func (f *fakeUI) Sound(prompt string)                    {}
func (f *fakeUI) StateChange(desc string)                { f.log = append(f.log, "stato:"+desc) }
func (f *fakeUI) Notice(text string)                     {}
func (f *fakeUI) Problem(text string)                    { f.problems = append(f.problems, text) }
func (f *fakeUI) Finish(o engine.Outcome)                {}

func (f *fakeUI) ChooseAction(p engine.ActionPrompt) (engine.Command, error) {
	for _, h := range p.Hidden {
		f.hidden[h.Action.ID] = h.Reason
	}
	tok, ok := f.next()
	if !ok {
		return engine.Command{}, engine.ErrScriptEnded
	}
	want := strings.TrimPrefix(tok, "a:")
	for _, a := range p.Available {
		if a.ID == want {
			return engine.Command{ActionID: a.ID}, nil
		}
	}
	f.problems = append(f.problems, "test: azione "+want+" non disponibile in "+p.Scene.ID)
	return engine.Command{Quit: true}, nil
}

func (f *fakeUI) ChooseChoice(p engine.ChoicePrompt) (engine.Command, error) {
	tok, ok := f.next()
	if !ok {
		return engine.Command{}, engine.ErrScriptEnded
	}
	want := strings.TrimPrefix(tok, "c:")
	for i, c := range p.Available {
		if c.Goto == want {
			return engine.Command{ChoiceIndex: i}, nil
		}
	}
	f.problems = append(f.problems, "test: scelta "+want+" non disponibile nel nodo "+p.NodeID)
	return engine.Command{Quit: true}, nil
}

func load(t *testing.T, path string) *ir.Story {
	t.Helper()
	s, err := ir.Load(path)
	if err != nil {
		t.Fatalf("caricamento %s: %v", path, err)
	}
	return s
}

func TestPartitaCompleta(t *testing.T) {
	story := load(t, "../../testdata/mini.ir.json")
	ui := newFake("a:continua", "a:parla_oste", "c:d_chiave", "a:prendi_chiave", "a:esci")
	e := engine.New(story, ui)
	out := e.Run()

	if len(out.Problems) > 0 {
		t.Fatalf("nessun problema atteso, trovati: %v", out.Problems)
	}
	if !out.Ended {
		t.Fatalf("la storia doveva arrivare a un finale, invece: %s", out.Reason)
	}
	if out.Scene != "finale" {
		t.Errorf("scena finale attesa 'finale', trovata %q", out.Scene)
	}
	if !e.State.HasItem("chiave") {
		t.Errorf("la chiave doveva essere in inventario: %v", e.State.Inventory)
	}
	if !e.State.Flags["oste_parlato"] || !e.State.Flags["prologo_visto"] {
		t.Errorf("flag attesi mancanti: %v", e.State.SortedFlags())
	}
	want := []string{"prologo", "taverna", "finale"}
	if strings.Join(e.State.History, ",") != strings.Join(want, ",") {
		t.Errorf("storico atteso %v, trovato %v", want, e.State.History)
	}
	// La traccia deve essere rigiocabile tale e quale.
	wantTrace := "a:continua,a:parla_oste,c:d_chiave,a:prendi_chiave,a:esci"
	if got := strings.Join(out.Trace, ","); got != wantTrace {
		t.Errorf("traccia attesa %q, trovata %q", wantTrace, got)
	}
}

func TestAzioneNascostaFinoAlFlag(t *testing.T) {
	story := load(t, "../../testdata/mini.ir.json")
	// Si tenta di prendere la chiave prima di parlare con l'oste.
	ui := newFake("a:continua", "a:prendi_chiave")
	engine.New(story, ui).Run()

	reason, ok := ui.hidden["prendi_chiave"]
	if !ok {
		t.Fatalf("prendi_chiave doveva risultare nascosta, nascoste: %v", ui.hidden)
	}
	if !strings.Contains(reason, "oste_parlato") {
		t.Errorf("il motivo doveva citare il flag mancante, invece: %q", reason)
	}
	if len(ui.problems) == 0 {
		t.Errorf("il tentativo su un'azione non disponibile doveva essere segnalato")
	}
}

func TestAzioneNonRipetibileSiConsuma(t *testing.T) {
	story := load(t, "../../testdata/mini.ir.json")
	ui := newFake("a:continua", "a:parla_oste", "c:d_chiave", "a:prendi_chiave", "a:prendi_chiave")
	engine.New(story, ui).Run()

	if reason := ui.hidden["prendi_chiave"]; !strings.Contains(reason, "repeatable") {
		t.Errorf("dopo l'uso l'azione doveva essere consumata, motivo: %q", reason)
	}
}

func TestScenaTerminaleChiudeLoScript(t *testing.T) {
	story := load(t, "../../testdata/mini.ir.json")
	// Lo script si ferma appena entrati nel finale: non e' un fallimento.
	ui := newFake("a:continua", "a:parla_oste", "c:d_chiave", "a:prendi_chiave", "a:esci")
	out := engine.New(story, ui).Run()
	if !out.Ended || out.Quit {
		t.Errorf("il finale doveva chiudere la partita, invece: %+v", out)
	}
}

func TestEffettiApplicatiInOrdine(t *testing.T) {
	st := engine.NewState()
	ui := newFake()
	tr := st.Apply(&ir.Effect{
		Narration:    "testo",
		SetFlag:      "a",
		AddInventory: "oggetto",
		GotoScene:    "altrove",
	}, ui)

	if tr.Kind != engine.TransScene || tr.Target != "altrove" {
		t.Errorf("transizione di scena attesa, trovata %+v", tr)
	}
	if !st.Flags["a"] || !st.HasItem("oggetto") {
		t.Errorf("stato non aggiornato: %v %v", st.SortedFlags(), st.Inventory)
	}
	if len(ui.log) == 0 || ui.log[0] != "narr" {
		t.Errorf("la narrazione doveva precedere i cambi di stato, log: %v", ui.log)
	}
}

func TestGotoSceneInesistenteEUnProblema(t *testing.T) {
	story := load(t, "../../testdata/rotta.ir.json")
	ui := newFake("a:porta_rotta")
	out := engine.New(story, ui).Run()

	if len(out.Problems) == 0 {
		t.Fatalf("un goto verso una scena inesistente doveva essere segnalato")
	}
	if !strings.Contains(out.Problems[0], "scena_inesistente") {
		t.Errorf("il problema doveva citare la scena mancante: %q", out.Problems[0])
	}
}

File: player/test/engine.test.ts
/**
 * Test dell'engine, portati uno a uno da quelli del player Go: se il
 * comportamento fosse cambiato nella migrazione, e' qui che si vede.
 */

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

import {
  Engine,
  GameState,
  ScriptEndedError,
  parseStory,
  type ActionPrompt,
  type ChoicePrompt,
  type Command,
  type DialogueNode,
  type NarrationBeat,
  type Outcome,
  type PlayerUI,
  type Scene,
  type Story,
} from '../src/core/index.js';

function fixture(name: string): Story {
  const path = fileURLToPath(new URL(`../../testdata/${name}`, import.meta.url));
  return parseStory(readFileSync(path, 'utf8'));
}

/**
 * UI finta che gioca una sequenza di token predefinita e registra quello che
 * vede. E' la stessa idea dello script di playthrough, ridotta al minimo.
 */
class FakeUI implements PlayerUI {
  pos = 0;
  problems: string[] = [];
  log: string[] = [];
  hidden = new Map<string, string>();

  constructor(private toks: string[] = []) {}

  private next(): string {
    if (this.pos >= this.toks.length) throw new ScriptEndedError();
    return this.toks[this.pos++];
  }

  sceneEnter(_st: GameState, sc: Scene): void {
    this.log.push('scena:' + sc.id);
  }
  beat(sc: Scene, _b: NarrationBeat, _i: number, _n: number): void {
    this.log.push('beat:' + sc.id);
  }
  line(_sc: Scene, nodeId: string, _n: DialogueNode): void {
    this.log.push('nodo:' + nodeId);
  }
  narration(): void {
    this.log.push('narr');
  }
  sound(): void {}
  stateChange(desc: string): void {
    this.log.push('stato:' + desc);
  }
  notice(): void {}
  problem(text: string): void {
    this.problems.push(text);
  }
  finish(_o: Outcome): void {}

  async chooseAction(p: ActionPrompt): Promise<Command> {
    for (const h of p.hidden) this.hidden.set(h.action.id, h.reason);
    const want = this.next().replace(/^a:/, '');
    const a = p.available.find((x) => x.id === want);
    if (a) return { actionId: a.id };
    this.problems.push(`test: azione ${want} non disponibile in ${p.scene.id}`);
    return { quit: true };
  }

  async chooseChoice(p: ChoicePrompt): Promise<Command> {
    const want = this.next().replace(/^c:/, '');
    const i = p.available.findIndex((c) => c.goto === want);
    if (i >= 0) return { choiceIndex: i };
    this.problems.push(`test: scelta ${want} non disponibile nel nodo ${p.nodeId}`);
    return { quit: true };
  }
}

test('una partita completa arriva al finale senza problemi', async () => {
  const story = fixture('mini.ir.json');
  const ui = new FakeUI(['a:continua', 'a:parla_oste', 'c:d_chiave', 'a:prendi_chiave', 'a:esci']);
  const e = new Engine(story, ui);
  const out = await e.run();

  assert.deepEqual(out.problems, [], 'nessun problema atteso');
  assert.ok(out.ended, `la storia doveva arrivare a un finale, invece: ${out.reason}`);
  assert.equal(out.scene, 'finale');
  assert.ok(e.state.hasItem('chiave'), `la chiave doveva essere in inventario: ${e.state.inventory}`);
  assert.ok(e.state.flags.has('oste_parlato') && e.state.flags.has('prologo_visto'));
  assert.deepEqual(e.state.history, ['prologo', 'taverna', 'finale']);
  // La traccia deve essere rigiocabile tale e quale.
  assert.deepEqual(out.trace, ['a:continua', 'a:parla_oste', 'c:d_chiave', 'a:prendi_chiave', 'a:esci']);
});

test("un'azione condizionata resta nascosta finche' manca il flag", async () => {
  const story = fixture('mini.ir.json');
  // Si tenta di prendere la chiave prima di parlare con l'oste.
  const ui = new FakeUI(['a:continua', 'a:prendi_chiave']);
  await new Engine(story, ui).run();

  const reason = ui.hidden.get('prendi_chiave');
  assert.ok(reason, `prendi_chiave doveva risultare nascosta, nascoste: ${[...ui.hidden.keys()]}`);
  assert.match(reason, /oste_parlato/, 'il motivo doveva citare il flag mancante');
  assert.ok(ui.problems.length > 0, "il tentativo su un'azione non disponibile doveva essere segnalato");
});

test("un'azione non ripetibile si consuma dopo il primo uso", async () => {
  const story = fixture('mini.ir.json');
  const ui = new FakeUI(['a:continua', 'a:parla_oste', 'c:d_chiave', 'a:prendi_chiave', 'a:prendi_chiave']);
  await new Engine(story, ui).run();

  assert.match(ui.hidden.get('prendi_chiave') ?? '', /repeatable/);
});

test('una scena terminale chiude la partita anche se lo script finisce', async () => {
  const story = fixture('mini.ir.json');
  const ui = new FakeUI(['a:continua', 'a:parla_oste', 'c:d_chiave', 'a:prendi_chiave', 'a:esci']);
  const out = await new Engine(story, ui).run();
  assert.ok(out.ended && !out.quit, `il finale doveva chiudere la partita: ${JSON.stringify(out)}`);
});

test("gli effetti si applicano nell'ordine dello schema", () => {
  const st = new GameState();
  const ui = new FakeUI();
  const tr = st.apply(
    { narration: 'testo', set_flag: 'a', add_inventory: 'oggetto', goto_scene: 'altrove' },
    ui,
  );

  assert.deepEqual(tr, { kind: 'scene', target: 'altrove' });
  assert.ok(st.flags.has('a') && st.hasItem('oggetto'));
  assert.equal(ui.log[0], 'narr', 'la narrazione doveva precedere i cambi di stato');
});

test('un goto verso una scena inesistente e\' un problema segnalato', async () => {
  const story = fixture('rotta.ir.json');
  const ui = new FakeUI(['a:porta_rotta']);
  const out = await new Engine(story, ui).run();

  assert.ok(out.problems.length > 0, 'un goto verso una scena inesistente doveva essere segnalato');
  assert.match(out.problems[0], /scena_inesistente/);
});

test('initial_inventory e\' gia\' in inventario prima della prima scena', async () => {
  // Un oggetto che il personaggio si porta dietro da prima della storia: la
  // sua unica azione lo richiede, e deve essere disponibile subito.
  const story: Story = {
    ir_version: '1.5.0',
    id: 'zaino',
    title: 'Con qualcosa nello zaino',
    start_scene: 'unica',
    items: [{ id: 'walkie_talkie', name: 'walkie talkie', aliases: ['radio', 'ricetrasmittente'] }],
    initial_inventory: ['walkie_talkie'],
    scenes: [
      {
        id: 'unica',
        background: { image_prompt: 'una stanza' },
        actions: [
          {
            id: 'usa_radio',
            label: 'Accendi la radio',
            condition: { has_item: 'walkie_talkie' },
            effect: { narration: 'Statica.' },
          },
        ],
      },
    ],
  };

  const ui = new FakeUI(['a:usa_radio']);
  const e = new Engine(story, ui);
  await e.run();

  assert.ok(e.state.hasItem('walkie_talkie'), 'la radio doveva essere in inventario dall\'inizio');
  assert.equal(ui.hidden.size, 0, "l'azione condizionata all'oggetto non doveva risultare nascosta");
});
